### ISSUE TITLE
PyTorch not reproducible on CPU

### ISSUE THREAD

**Issue by alanhdu** [pytorch/pytorch#32536]
## 🐛 Bug

We are currently using PyTorch in a real-time streaming situation, and we have a set of tests to ensure that a model's behavior in the online inference (where we get data points batch-by-batch) matches the behavior in offline inference (where we get the entire time series all at once). For reasons I can't understand though, we cannot get the two behaviors to line up.

## To Reproduce

If you run:

```python
import torch
import torch.nn as nn

torch.manual_seed(512)
data = torch.rand(2, 1)
module = nn.Linear(1, 2)

with torch.no_grad():
    batch_output = module.forward(data)

    outputs = [
        module.forward(data[0:1, ...]),
        module.forward(data[1:2, ...]),
    ]
    stream_output = torch.cat(outputs, dim=0)

assert torch.equal(batch_output, stream_output)
```

The assertion will fail. This is a little strange though -- `nn.Linear` is totally stateless and should just correspond to a `weights @ data + bias` based on the last axis of `data`, and so AFAICT these should be identical computations.

We've worked around this by using  `torch.allclose(..., atol=2e-7)` in our tests, but every so often I have to keep bumping up the `atol` to keep tests passing as our models get more complex.

## Expected behavior

The two Tensors produced would be *exactly* equal and reproducible.

## Environment

 - PyTorch Version (e.g., 1.0): 1.4 CPU and 1.3 CPU
 - OS (e.g., Linux): Linux (Ubuntu 18.04)
 - How you installed PyTorch (`conda`, `pip`, source): `conda`
 - Python version: 3.6.7

**Comment by albanD** [comment]
Hi,

Did you read the reproducibility note from the doc? https://pytorch.org/docs/stable/notes/randomness.html#reproducibility
Does that answer your question?

**Comment by alanhdu** [comment]
I did read it, but I don't think it's quite getting at the same thing: that's about reproducibility across repeated invocation of Pytorch, while this is about different types of invocations that should be mathematically equivalent.

That said, based on some in-person conversations with some other (more knowledgable) people, it seems like Pytorch is choosing different algorithms under-the-hood based on the size of the tensors, and so we do expect some amount of floating point error between the algorithms.
